src/realm/sync/apply_to_state_command.cpp

```diff
@@ -0,0 +1,407 @@
+#include <algorithm>
+#include <cinttypes>
+#include <string>
+#include <type_traits>
+
+#include <external/mpark/variant.hpp>
+
+#include "realm/db.hpp"
+#include "realm/sync/history.hpp"
+#include "realm/sync/instruction_applier.hpp"
+#include "realm/sync/impl/clamped_hex_dump.hpp"
+#include "realm/sync/noinst/client_history_impl.hpp"
+#include "realm/sync/protocol.hpp"
+#include "realm/sync/transform.hpp"
+#include "realm/sync/changeset_parser.hpp"
+#include "realm/sync/noinst/compression.hpp"
+#include "realm/util/cli_args.hpp"
+#include "realm/util/load_file.hpp"
+#include "realm/util/safe_int_ops.hpp"
+#include "realm/util/string_view.hpp"
+
+using namespace realm::util;
+
+template <typename T>
+using ParseResult = std::pair<T, StringView>;
+
+struct ServerIdentMessage {
+    realm::sync::session_ident_type session_ident;
+    realm::sync::SaltedFileIdent file_ident;
+
+    static ParseResult<ServerIdentMessage> parse(StringView sv);
+};
+
+struct DownloadMessage {
+    realm::sync::session_ident_type session_ident;
+    realm::sync::SyncProgress progress;
+    realm::sync::SaltedVersion latest_server_version;
+    uint64_t downloadable_bytes;
+
+    Buffer<char> uncompressed_body_buffer;
+    std::vector<realm::sync::Transformer::RemoteChangeset> changesets;
+
+    static ParseResult<DownloadMessage> parse(StringView sv, Logger* logger);
+};
+
+struct UploadMessage {
+    realm::sync::session_ident_type session_ident;
+    realm::sync::UploadCursor upload_progress;
+    realm::sync::version_type locked_server_version;
+
+    Buffer<char> uncompressed_body_buffer;
+    std::vector<realm::sync::Changeset> changesets;
+
+    static ParseResult<UploadMessage> parse(StringView sv, Logger* logger);
+};
+
+using Message = mpark::variant<ServerIdentMessage, DownloadMessage, UploadMessage>;
+
+struct MessageParseException : public std::runtime_error {
+    using std::runtime_error::runtime_error;
+
+    template <typename... Args>
+    MessageParseException(const char* fmt, Args&&... args)
+        : std::runtime_error(format(fmt, args...))
+    {
+    }
+};
+
+// These functions will parse the space/new-line delimited headers found at the beginning of
+// messages and changesets.
+StringView parse_header_element(StringView sv, char)
+{
+    return sv;
+}
+
+template <typename T, typename = std::enable_if_t<std::is_integral_v<std::remove_reference_t<T>>>>
+StringView parse_header_value(StringView sv, T&& cur_arg)
+{
+    char* end_ptr = nullptr;
+    if constexpr (std::is_signed_v<T>) {
+        auto big_val = std::strtoll(sv.data(), &end_ptr, 10);
```

Comment:
Unfortunately, `strtoll` is locale-sensitive. Not sure if we care enough, but you can do use `std::stringstream` instead if you want.

```
std::istringstream ss{sv};
ss.imbue(std::locale::classic());
ss >> big_val;
```

Or alternatively call `std::locale::global(std::locale::classic())` when the program begins (this also sets the C locale).